test/test-tcp-reuseport.c

```diff
@@ -0,0 +1,191 @@
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#include "uv.h"
+#include "task.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#if !defined(__linux__) && !defined(__FreeBSD__) && !defined(__DragonFly__) && !defined(__sun) && !defined(_AIX73)
+
+TEST_IMPL(tcp_reuseport) {
+  RETURN_SKIP("Run this test only on Linux, FreeBSD, DragonFlyBSD, Solaris, and AIX for now.");
+}
+
+#else
+
+#define MAX_TCP_CLIENTS 10
+
+static uv_tcp_t tcp_connect_handles[MAX_TCP_CLIENTS];
+static uv_connect_t tcp_connect_requests[MAX_TCP_CLIENTS];
+
+static unsigned int main_loop_accepted;
+static unsigned int thread_loop_accepted;
+static unsigned int connected;
+
+static uv_mutex_t mutex;
+static unsigned int accepted;
+
+static uv_loop_t* main_loop;
+static uv_loop_t* thread_loop;
+static uv_tcp_t main_handle;
+static uv_tcp_t thread_handle;
+static uv_timer_t main_timer_handle;
+static uv_timer_t thread_timer_handle;
+
+static void on_close(uv_handle_t* handle) {
+  free(handle);
+}
+
+static void ticktack(uv_timer_t* timer) {
+  ASSERT(timer == &main_timer_handle || timer == &thread_timer_handle);
+
+  int done = 0;
+  uv_mutex_lock(&mutex);
+  if (accepted == MAX_TCP_CLIENTS) {
+    done = 1;
+  }
+  uv_mutex_unlock(&mutex);
+
+  if (done) {
+    uv_close((uv_handle_t*) timer, NULL);
+    if (timer->loop == main_loop)
+      uv_close((uv_handle_t*) &main_handle, NULL);
+    if (timer->loop == thread_loop)
+      uv_close((uv_handle_t*) &thread_handle, NULL);
+  }
+}
+
+static void on_connection(uv_stream_t* server, int status)
+{
+  ASSERT_OK(status);
+  ASSERT(server == (uv_stream_t*) &main_handle || server == (uv_stream_t*) &thread_handle);
+
+  uv_tcp_t *client = malloc(sizeof(uv_tcp_t));
+  ASSERT_OK(uv_tcp_init(server->loop, client));
+  ASSERT_OK(uv_accept(server, (uv_stream_t*) client));
+  uv_close((uv_handle_t*) client, on_close);
+
+  if (server->loop == main_loop)
+    main_loop_accepted++;
+
+  if (server->loop == thread_loop)
+    thread_loop_accepted++;
+
+  uv_mutex_lock(&mutex);
+  accepted++;
```

Comment:
You are not asserting on this value, are you? Maybe add an assert too.

Reply:
It's actually "asserted" in function `ticktack` for `MAX_TCP_CLIENTS` to ensure `main_loop` and `thread_loop` would return from `uv_run()`, otherwise this test would fail with a timeout. But we can double-assert it, it never hurts to be cautious.